tlatools/org.lamport.tlatools/src/tla2sany/semantic/SemanticNode.java

```diff
@@ -301,6 +301,52 @@ public void walkGraph(Hashtable<Integer, ExploreNode> semNodesTable, ExplorerVis
     visitor.preVisit(this);
     visitor.postVisit(this);
   }
+  
+  public boolean isDefinedWith(final SemanticNode sn) {
+	final ExplorerVisitor<Boolean> visitor = new ExplorerVisitor<Boolean>() {
+		private boolean cycle = false;
+		private boolean substInNode = false;
+
+		@Override
+		public void preVisit(ExploreNode exploreNode) {
+			if (exploreNode instanceof SubstInNode) {
+				//TODO This is nothing but a bandaid.
+				substInNode = true;
+			}
+			if (exploreNode == sn) {
+				cycle = true;
+			}
+		}
+			
+		public Boolean get() {
+			return cycle && !substInNode;
+		}
+	};
+	this.walkGraph(new Hashtable<>(), visitor);
+	return visitor.get();
+  }
+
+  // Replaces all occurrences of OpDefNode o with OpDefNode s.
+  public void substituteFor(final OpDefNode s, final OpDefNode o) {
+	// Traverse the semantic graph starting at its root node and find all
+	// applications (OpApplNode) of o and replace them with s, except the
+	// applications of o that appear in the definition of s.
+	// TODO Is there a cheaper way to find all applications (OpApplNode) of a
+    // definition (OpDefNode)?
+	final ExplorerVisitor<Void> explorerVisitor = new ExplorerVisitor<Void>() {
+		public void preVisit(final ExploreNode en) {
+			if (en instanceof OpApplNode) {
+				final OpApplNode oan = (OpApplNode) en;
+				if (oan.getOperator() == o) {
+					if (!s.isDefinedWith(oan)) {
+						oan.resetOperator(s);
```

Comment:
The changes in this PR look reasonable---although this specific line looks a little fishy to me because it modifies the semantic graph _while the graph is being traversed_. There is a good reason the Java standard collections forbid this kind of behavior, and those data structures are much simpler than this one.

If we do end up merging this, I would want to at least document what kinds of changes `preVisit` and `postVisit` are allowed to make.

Reply:
https://github.com/tlaplus/tlaplus/pull/825#issuecomment-1676620517